x-pack/plugin/old-lucene-versions/src/main/java/org/elasticsearch/xpack/lucene/bwc/OldLuceneVersions.java

```diff
@@ -201,6 +202,12 @@ private static SegmentInfos convertToNewerLuceneVersion(OldSegmentInfos oldSegme
         if (map.containsKey(Engine.MAX_UNSAFE_AUTO_ID_TIMESTAMP_COMMIT_ID) == false) {
             map.put(Engine.MAX_UNSAFE_AUTO_ID_TIMESTAMP_COMMIT_ID, "-1");
         }
+        if (map.containsKey(Engine.ES_VERSION) == false) {
+            assert oldSegmentInfos.getLuceneVersion()
+                .onOrAfter(RecoverySettings.SEQ_NO_SNAPSHOT_RECOVERIES_SUPPORTED_VERSION.luceneVersion()) == false
+                : oldSegmentInfos.getLuceneVersion() + " should contain the ES_VERSION";
+            map.put(Engine.ES_VERSION, IndexVersion.current().toString());
```

Comment:
Not 100% sure I understand why we had to add this, but also does not seem like a big issue.

I do wonder if we want to use the N-1 version like we do for the compatibility version?

Reply:
While I am aware that ES_VERSION is a setting that old lucene versions add as part of manipulating index settings upon restore/mount, I also don't follow why this change was required as part of supporting N-2. I thought that we made sure that in main, archive indices only target indices before N-2, for which this setting is already added? Maybe it's a matter of sequencing, is it now requested earlier?

Reply:
> Not 100% sure I understand why we had to add this, but also does not seem like a big issue.

The `ES_VERSION` was added for snapshpot-based recoveries to ensure we can use a snapshot instead of peer-recovery. An assertion was also added in `IndexShard.assertSequenceNumbersInCommit` (now renamed to `assertLastestCommitUserData`) to ensure `ES_VERSION` exists in the Lucene commit.

For old lucene versions mounted as searchable snapshots, the ES_VERSION was transparently added by the Lucene commit executed to associate the empty translog with the Lucene index. The list of commit user data was just reused at it is by the OldLuceneVersion plugin to rewrite the segment infos. 

Now we don't execute any commit during the searchable snapshot recovery, the ES_VERSION is missing from the Lucene commit. I think it is correct that the OldLuceneVersion plugin adds this information if it does not exist.

> I also don't follow why this change was required as part of supporting N-2. I thought that we made sure that in main, archive indices only target indices before N-2, for which this setting is already added? Maybe it's a matter of sequencing, is it now requested earlier?

For supporting N-2 indices we need to make sure that we are not writing those indices. This is the purpose of this pull request, ie not commiting the Lucene index to associate a translog that is unnecessary. By skipping this commit, the ES_VERSION is not guarantee to be present for Lucene indices before 7.15, so we need to add this information back (because elasticsearch expects it, at least for an assertion).

> I do wonder if we want to use the N-1 version like we do for the compatibility version?

Good point, I changed that.